Disallow generic type parameters from appearing within certain constants (#68356)
rust-lang/rust: src/librustc_error_codes/error_codes/E0747.md

> Any constant items, static items, array length expressions, or enum

Reviewer: ```suggestion
Any constant item, static item, array length expression, or enum
```

> discriminants can't refer to type parameters.

Reviewer: ```suggestion
discriminant cannot refer to type parameters.
```

Erroneous code example:

```compile_fail,E0747
use std::mem::size_of;

fn foo<T>() {
    let _ = [0; size_of::<T>()];
}
```

A workaround for array length expressions is to use [`vec!`] instead, which
does support type parameters in it's length expression, though this does
perform dynamic memory allocation. Example:

```
use std::mem::size_of;

fn foo<T>() {
    let _ = vec![0; size_of::<T>()];
}
```

> Though enum discriminants can't refer to regular type parameters they can still

Reviewer: ```suggestion
While enum discriminants cannot refer to regular type parameters, they can still
```

refer to the `Self` type parameter. Example:

```
#[repr(u8)]
enum Alpha {
    V1 = 41,
    V2 = Self::V1 as u8 + 1,
}
```

> Note, associated constants do not have this limitation and they can refer to

Reviewer: ```suggestion
Note that associated constants do not have this limitation and can refer to
```

type parameters. Example:

```
use std::mem::size_of;

trait Foo {
    const X: i32;
}

struct Bar<T>(T);

impl<T> Foo for Bar<T> {
    const X: i32 = size_of::<T>();
}
```

> This is currently a limitation with the compiler and these restrictions may be

Reviewer: ```suggestion
This is currently a limitation with the compiler. These restrictions may be
```
